conda/gateways/repodata/__init__.py

```diff
@@ -663,6 +698,246 @@ def timeout(self):
         return ((now - refresh) + max_age) / 1e9
 
 
+class RepodataFetch:
+    """
+    Combine RepodataCache and RepoInterface to provide subdir_data.SubdirData()
+    with what it needs.
+
+    Provide a variety of formats since some ``RepoInterface`` have to
+    ``json.loads(...)`` anyway, and some clients don't need the Python data
+    structure at all.
+    """
+
+    cache_path_base: Path
+    channel: Channel
+    repodata_fn: str
+    url_w_subdir: str
+    url_w_credentials: str
+    repo_interface_cls: Any
+
+    def __init__(
+        self, cache_path_base: Path, channel: Channel, repodata_fn: str, *, repo_interface_cls
+    ):
+        self.cache_path_base = cache_path_base
+        self.channel = channel
+        self.repodata_fn = repodata_fn
+
+        self.url_w_subdir = self.channel.url(with_credentials=False) or ""
+        self.url_w_credentials = self.channel.url(with_credentials=True) or ""
+
+        self.repo_interface_cls = repo_interface_cls
+
+    def fetch_latest_parsed(self) -> tuple[dict, RepodataState]:
+        """
+        Retrieve parsed latest or latest-cached repodata as a dict; update
+        cache.
+
+        :return: (repodata contents, state including cache headers)
+        """
+        parsed, state = self.whatever_subdir_data_used_to_do()
+        if isinstance(parsed, str):
+            return json.loads(parsed), state
+        else:
+            return parsed, state
+
+    def fetch_latest_str(self) -> tuple[str, RepodataState]:
```

Comment:
It might be possible to get rid of the `_str` variant. SubdirData needs the json, and libmamba-solver only needs to be able to load it again from disk.

It is very awkward to change the functions below this, that have to download and return the index data, to either return str, parsed json or a Path.